api/provider/aws/deploy_create.go

```diff
@@ -0,0 +1,136 @@
+// and calls the AWS ECS API with it. In order to do this, we must:
+//
+// - Generate a unique deployID and fqDeployID
+// - Render a AWS TaskDefinition from the Request
+// - Create an ecs.TaskDefinition object for the Task Definition data
+// - Call the AWS' ECS RegisterTaskDefinition API with the previous steps' object
+// - If the AWS API call is successful, Marshal the request into a []byte
+//   - Instantiate Layer0 models.Deploy object with the marshaled data
+// - Insert entity tags into Layer0 tags database
+// - Return entity object back to Controller layer
+//
+func (d *DeployProvider) Create(req models.CreateDeployRequest) (*models.Deploy, error) {
+	deployID := generateEntityID(req.DeployName)
+	fqDeployID := addLayer0Prefix(d.Config.Instance(), deployID)
+	familyName := fqDeployID
+
+	taskDefinition, err := d.renderTaskDefinition(req.DeployFile, fqDeployID)
+	if err != nil {
+		return nil, err
+	}
+
+	input := &ecs.TaskDefinition{}
+	input.SetFamily(familyName)
+	input.SetTaskRoleArn(aws.StringValue(taskDefinition.TaskRoleArn))
+	input.SetNetworkMode(aws.StringValue(taskDefinition.NetworkMode))
+	input.SetContainerDefinitions(taskDefinition.ContainerDefinitions)
+	input.SetVolumes(taskDefinition.Volumes)
+	input.SetPlacementConstraints(taskDefinition.PlacementConstraints)
+
+	taskDefinitionOutput, err := d.createTaskDefinition(input)
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := json.Marshal(taskDefinitionOutput)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to extract deploy file: %s", err.Error())
+	}
+
+	deploy := &models.Deploy{
+		// TODO: we previously removed prefix from deployID ala below. Is this
+		// no longer needed?
+		// func removePrefix(id string) string {
+		// 	return strings.TrimPrefix(id, PREFIX)
+		// }
+		DeployID:   deployID,
+		Version:    d.getDeployRevision(deployID),
+		DeployFile: bytes,
+	}
+
+	if err := d.TagStore.Insert(models.Tag{EntityID: deploy.DeployID, EntityType: "deploy", Key: "name", Value: req.DeployName}); err != nil {
+		return deploy, err
+	}
+
+	if err := d.TagStore.Insert(models.Tag{EntityID: deploy.DeployID, EntityType: "deploy", Key: "version", Value: deploy.Version}); err != nil {
+		return deploy, err
+	}
+
+	return deploy, nil
+}
+
+func (d *DeployProvider) createTaskDefinition(taskDefinitionRequest *ecs.TaskDefinition) (*ecs.TaskDefinition, error) {
+
+	input := &ecs.RegisterTaskDefinitionInput{}
+	input.SetFamily(aws.StringValue(taskDefinitionRequest.Family))
+	input.SetTaskRoleArn(aws.StringValue(taskDefinitionRequest.TaskRoleArn))
+	input.SetNetworkMode(aws.StringValue(taskDefinitionRequest.NetworkMode))
+	input.SetContainerDefinitions(taskDefinitionRequest.ContainerDefinitions)
+	input.SetVolumes(taskDefinitionRequest.Volumes)
+	input.SetPlacementConstraints(taskDefinitionRequest.PlacementConstraints)
+
+	if err := input.Validate(); err != nil {
+		return nil, err
+	}
+
+	_, err := d.AWS.ECS.RegisterTaskDefinition(input)
```

Comment:
```
output, err := d.AWS.ECS.RegisterTaskDefinition(input)
if err != nil {
    return nil, err
}

return output.TaskDefinition, nil
```

Reply:
I'm fine with doing this, but it isn't quite consistent with what's in `environment`, and I think consistency is something really critical for this package. That said, with your `environment` helper functions, you aren't passing AWS objects around like I am... not sure if passing those objects is the best way forward here or not.

Reply:
What's not consistent about it? 

We pass in `aws-sdk-go` objects into helpers when it makes the most sense (see [CreateLoadBalancer](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/load_balancer_create.go#L148)). We also return `aws-sdk-go` objects from helper functions when it makes the most sense (see [ReadEnvironment](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/environment_read.go#L48)).